source/validate_cfg.cpp

```diff
@@ -24,32 +24,314 @@
 // TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 // MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
 
+#include "validate.h"
 #include "validate_passes.h"
 
+#include <algorithm>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+using std::find;
+using std::get;
+using std::make_pair;
+using std::make_tuple;
+using std::numeric_limits;
+using std::pair;
+using std::transform;
+using std::tuple;
+using std::unordered_map;
+using std::vector;
+
+using libspirv::BasicBlock;
+
 namespace libspirv {
 
-// TODO(umar): Support for merge instructions
-// TODO(umar): Structured control flow checks
+namespace {
+
+using bb_ptr = BasicBlock*;
+using cbb_ptr = const BasicBlock*;
+using bb_iter = vector<BasicBlock*>::const_iterator;
+
+template <typename T>
+bool Contains(const vector<T>& vec, T val) {
+  return find(begin(vec), end(vec), val) != end(vec);
+}
+
+/// @brief Sorts the blocks in a CFG given the entry node
+///
+/// Returns a vector of basic block pointers in a Control Flow Graph(CFG) which
+/// are sorted in the order they were accessed in a post order traversal.
+///
+/// @param[in] entry the first block of a CFG
+/// @param[in] depth_hint a hint about the depth of the CFG
+///
+/// @return A vector of pointers in the order they were access in a post order
+/// traversal
+vector<const BasicBlock*> PostOrderSort(const BasicBlock& entry, size_t size) {
```

Comment:
I was trying to avoid recursion as we don't know how deep the user's code could nest. I guess we can rely on the compiler to perform tail call optimization. I am not partial to either approach. Your call.